### PR TITLE
Report unknown licenses  separately.

### LINKED ISSUE #2574: Unknown Licenses should be reported separately
We should have a separate section for reporting unknown licenses and not mixed with all license detections for clarity.

## 

- new feature

## Select Category

-  [x] Enhancement
-  [ ] Add License/Copyright
-  [x] Scan Feature
-  [ ] Packaging
-  [ ] Documentation
-  [ ] Expand Support
-  [ ] Other

## **Possible Solution/Implementation Details**

#1675 for more details

### PR DESCRIPTION
**Pull request by akugarg**
Signed-off-by: akugarg <akanksha.garg2k@gmail.com>

Fixes #2574 

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

**Comment by AyanSinhaMahapatra** [comment]
Btw, we have to create a resource attribute like this [here](https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/plugin_license.py#L42), and you have to return `unknown_licenses` [here](https://github.com/nexB/scancode-toolkit/blob/develop/src/scancode/api.py#L201).

File: src/scancode/api.py [hunk 1]
```diff
@@ -181,15 +181,25 @@ def get_licenses(location, min_score=0,
         qspans.append(match.qspan)
 
         detected_expressions.append(match.rule.license_expression)
-
-        detected_licenses.extend(
-            _licenses_data_from_match(
+        if "unknown" in match.rule.license_expression:
```
Comment (AyanSinhaMahapatra) [inline src/scancode/api.py]: matches will have a `is_unknown` flag you added :P Use that instead of checking the license_eexpression.

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]
Btw, we have to create a resource attribute like this [here](https://github.com/nexB/scancode-toolkit/blob/develop/src/licensedcode/plugin_license.py#L42), and you have to return `unknown_licenses` [here](https://github.com/nexB/scancode-toolkit/blob/develop/src/scancode/api.py#L201).

File: src/scancode/api.py (hunk 1 above)
Comment (akugarg) [reply src/scancode/api.py]: Oh yeah! Actually that PR is not yet merged therefore changes were not reflected in this branch which was giving errors, that's why I have used this .

File: src/scancode/api.py [hunk 2]
```diff
@@ -181,15 +181,25 @@ def get_licenses(location, min_score=0,
...
                 match=match,
                 include_text=include_text,
                 license_text_diagnostics=license_text_diagnostics,
                 license_url_template=license_url_template)
-        )
-
+            )
+    if len(unknown_licenses) != 0:
+        result={}
+        result['Unknown_licenses'] = unknown_licenses
```
Comment (pombredanne) [inline src/scancode/api.py]: we always use all lowercase keys.

File: src/scancode/api.py [hunk 3]
```diff
@@ -180,16 +181,23 @@ def get_licenses(location, min_score=0,
...
         qspans.append(match.qspan)
 
-        detected_expressions.append(match.rule.license_expression)
-
-        detected_licenses.extend(
-            _licenses_data_from_match(
+        if "unknown" in match.rule.license_expression:  # TODO: use is_unknwon flag instead
+            unknown_expressions.append(match.rule.license_expression)
+            unknown_licenses.extend(_licenses_data_from_match(
+                match=match,
```
Comment (AyanSinhaMahapatra) [inline src/scancode/api.py:188]: Using `_licenses_data_from_match` is repeated, this can be done only once by assigning the output to an intermediate variable, and then later extending the lists with it inside the if-else block.

File: src/scancode/api.py [hunk 4]
```diff
@@ -171,7 +171,8 @@ def get_licenses(location, min_score=0,
 
     detected_licenses = []
     detected_expressions = []
-
+    unknown_expressions = []
+    unknown_licenses=[]
```
Comment (AyanSinhaMahapatra) [inline src/scancode/api.py:175]: Should have a space between the lists and getting the matches like before.

File: src/licensedcode/plugin_license.py [hunk 5]
```diff
@@ -41,7 +41,10 @@ class LicenseScanner(ScanPlugin):
 
     resource_attributes = dict([
         ('licenses', attr.ib(default=attr.Factory(list))),
+        ('unknown_licenses', attr.ib(default=attr.Factory(list))),
         ('license_expressions', attr.ib(default=attr.Factory(list))),
+        ('unknown_license_expressions', attr.ib(default=attr.Factory(list))),
+        ('spdx_license_expressions', attr.ib(default=attr.Factory(list))),
```
Comment (AyanSinhaMahapatra) [inline src/licensedcode/plugin_license.py]: Adding the `spdx_license_expressions` should be done seperately in a seperate PR (or atleast after all the tests pass for reporting unknowns seperately), to see the effects seperately.

File: src/scancode/api.py [hunk 6]
```diff
@@ -180,16 +181,23 @@ def get_licenses(location, min_score=0,
...
+            detected_licenses.extend(
+                 _licenses_data_from_match(
                 match=match,
                 include_text=include_text,
                 license_text_diagnostics=license_text_diagnostics,
                 license_url_template=license_url_template)
-        )
-
+            )
+
```
Comment (AyanSinhaMahapatra) [inline src/scancode/api.py:200]: Careful not to leave whitespace in empty lines.

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]
Should fix the tests next.

**Comment by AyanSinhaMahapatra** [comment]
Some tests are failing in `summarycode/plugin_consolidate.py`. See [here](https://github.com/nexB/scancode-toolkit/blob/develop/src/summarycode/plugin_consolidate.py#L336). Another `normalized_unknown_license_expression` should be saved to each child, and then added to [Consolidation.other_license_expression](https://github.com/nexB/scancode-toolkit/blob/develop/src/summarycode/plugin_consolidate.py#L61). And related test expectations fixed.

File: src/summarycode/plugin_consolidate.py [hunk 7]
```diff
@@ -383,15 +386,19 @@ def create_consolidated_components(resource, codebase, holder_key):
     given resource and holder key
     """
     license_expressions = []
+    unknwown_expressions = []
```
Comment (AyanSinhaMahapatra) [inline src/summarycode/plugin_consolidate.py]: `unknwown_expressions` should be `unknown_expressions`

Comment (akugarg) [reply src/summarycode/plugin_consolidate.py]: Oops! typo

File: src/summarycode/plugin_consolidate.py [hunk 8]
```diff
@@ -338,6 +338,9 @@ def get_holders_consolidated_components(codebase):
                 if license_expression:
                     child.extra_data['normalized_license_expression'] = license_expression
                     child.save(codebase)
+                if "unknown" in license_expression:
```
Comment (AyanSinhaMahapatra) [inline src/summarycode/plugin_consolidate.py]: The `license_expression` is added to both `normalized_license_expression` and `normalized_unknown_license_expression`. If `unknown` add only to `normalized_unknown_license_expression` and not in `normalized_license_expression`.

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]

File: src/summarycode/plugin_consolidate.py [hunk 9]
```diff
@@ -404,6 +412,7 @@ def create_consolidated_components(resource, codebase, holder_key):
 
     c = Consolidation(
         core_license_expression=combine_expressions(license_expressions),
+        other_license_expression = combine_expressions(unknown_expressions),
```
Comment (AyanSinhaMahapatra) [inline src/summarycode/plugin_consolidate.py]: No spaces should be present before and after `=` in this case.

File: src/scancode/api.py [hunk 10]
```diff
@@ -179,17 +181,23 @@ def get_licenses(location, min_score=0,
...
+        matched_license.extend(
             _licenses_data_from_match(
                 match=match,
                 include_text=include_text,
                 license_text_diagnostics=license_text_diagnostics,
                 license_url_template=license_url_template)
         )
-
+        
+        if "unknown" in match.rule.license_expression:  #TODO: use is_unknown flag instead
```
Comment (pombredanne) [inline src/scancode/api.py]: re `#TODO: use is_unknown flag instead` when would this happen?

File: src/scancode/api.py [hunk 11]
```diff
@@ -200,7 +208,9 @@ def get_licenses(location, min_score=0,
     detected_spdx_expressions = []
     return dict([
         ('licenses', detected_licenses),
+        ('unknown_licenses', unknown_licenses),
```
Comment (pombredanne) [inline src/scancode/api.py]: I would rather have licenses, license_expressions, unknown_licenses, unknown_license_expressions in this order everywhere.... this make more sense to group unknowns together.

File: tests/cluecode/data/plugin_filter_clues/filtered-expected2.json [hunk 12]
```diff
@@ -77,9 +77,11 @@
           }
         }
       ],
+      "unknown_licenses": [],
```
Comment (pombredanne) [inline tests/cluecode/data/plugin_filter_clues/filtered-expected2.json]: unknown_licenses are less important than licenses so they should come after always

File: tests/cluecode/data/plugin_filter_clues/filtered-expected3.json [hunk 13]
```diff
@@ -33,12 +33,12 @@
       "sha256": "408dfe8f9a70ea7bc9feaa9f77fb731ada0c26dc39008fdc5ead52d0442c656f",
       "mime_type": "text/plain",
       "file_type": "ASCII text",
-      "programming_language": null,
+      "programming_language": "verilog",
```
Comment (pombredanne) [inline tests/cluecode/data/plugin_filter_clues/filtered-expected3.json]: This is weird... how is this related to the current PR?

File: tests/formattedcode/data/json/tree/expected.json [hunk 14]
```diff
@@ -329,7 +345,10 @@
       "is_source": true,
       "is_script": false,
       "licenses": [],
+      "unknown_licenses": [],
       "license_expressions": [],
+      "unknown_license_expressions": [], "license_expressions": [],
+      "unknown_license_expressions": [],
```
Comment (pombredanne) [inline tests/formattedcode/data/json/tree/expected.json:348]: This is not right: you cannot have twice the same keys? This is 
```
      "unknown_licenses": [],
      "license_expressions": [],
      "unknown_license_expressions": [], "license_expressions": [],
      "unknown_license_expressions": [],

```

and the order should be:
```
      "licenses": [],
      "license_expressions": [],
      "unknown_licenses": [],
      "unknown_license_expressions": [],
```

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks. See my comments inline.

File: src/scancode/api.py (hunk 10 above)
Comment (AyanSinhaMahapatra) [reply src/scancode/api.py]: @pombredanne after https://github.com/nexB/scancode-toolkit/pull/2548 is approved and merged, then this could be rebased/merged and the `is_unknown` flags introduced there can be used.

**Comment by pombredanne** [comment]
@akugarg @AyanSinhaMahapatra should we merge this? or is this no longer relevant based on our latest discussions?

**Comment by AyanSinhaMahapatra** [comment]
@pombredanne Yes, this is not relevant anymore. @akugarg Thanks though.
